ur_controllers/src/freedrive_mode_controller.cpp

```diff
@@ -0,0 +1,330 @@
+  freedrive_param_listener_->refresh_dynamic_parameters();
+
+  // Get parameters from the listener in case they were updated
+  freedrive_params_ = freedrive_param_listener_->get_params();
+
+  start_logging_thread();
+
+  return ControllerInterface::on_configure(previous_state);
+}
+
+controller_interface::CallbackReturn
+ur_controllers::FreedriveModeController::on_activate(const rclcpp_lifecycle::State& state)
+{
+  change_requested_ = false;
+  freedrive_active_ = false;
+  async_state_ = std::numeric_limits<double>::quiet_NaN();
+
+  first_log_ = false;
+  logging_thread_running_ = false;
+  logging_requested_ = false;
+
+  {
+    const std::string interface_name = freedrive_params_.tf_prefix + "freedrive_mode/"
+                                                                     "async_success";
+    auto it = std::find_if(command_interfaces_.begin(), command_interfaces_.end(),
+                           [&](auto& interface) { return (interface.get_name() == interface_name); });
+    if (it != command_interfaces_.end()) {
+      async_success_command_interface_ = *it;
+    } else {
+      RCLCPP_ERROR(get_node()->get_logger(), "Did not find '%s' in command interfaces.", interface_name.c_str());
+      return controller_interface::CallbackReturn::ERROR;
+    }
+  }
+
+  {
+    const std::string interface_name = freedrive_params_.tf_prefix + "freedrive_mode/"
+                                                                     "enable";
+    auto it = std::find_if(command_interfaces_.begin(), command_interfaces_.end(),
+                           [&](auto& interface) { return (interface.get_name() == interface_name); });
+    if (it != command_interfaces_.end()) {
+      enable_command_interface_ = *it;
+    } else {
+      RCLCPP_ERROR(get_node()->get_logger(), "Did not find '%s' in command interfaces.", interface_name.c_str());
+      return controller_interface::CallbackReturn::ERROR;
+    }
+  }
+
+  {
+    const std::string interface_name = freedrive_params_.tf_prefix + "freedrive_mode/"
+                                                                     "abort";
+    auto it = std::find_if(command_interfaces_.begin(), command_interfaces_.end(),
+                           [&](auto& interface) { return (interface.get_name() == interface_name); });
+    if (it != command_interfaces_.end()) {
+      abort_command_interface_ = *it;
+      abort_command_interface_->get().set_value(0.0);
+    } else {
+      RCLCPP_ERROR(get_node()->get_logger(), "Did not find '%s' in command interfaces.", interface_name.c_str());
+      return controller_interface::CallbackReturn::ERROR;
+    }
+  }
+
+  return ControllerInterface::on_activate(state);
+}
+
+controller_interface::CallbackReturn
+ur_controllers::FreedriveModeController::on_deactivate(const rclcpp_lifecycle::State&)
+{
+  abort_command_interface_->get().set_value(1.0);
+  freedrive_active_ = false;
+
+  freedrive_sub_timer_.reset();
+  timer_started_ = false;
+
+  stop_logging_thread();
```

Comment:
This is definitively not RT-safe

Reply:
True, moved `stop_logging_thread()` to `on_cleanup()`. Thanks!